File: Tree/Lowest Common Ancestor of a Binary Tree/lowestCommonAncestor.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """
        pathP,pathQ = self.findPath(root,p),self.findPath(root,q)
        ans,x = None,0
        while x < min(len(pathP),len(pathQ)) and pathP[x] == pathQ[x]:
            ans,x = pathP[x],x+1
        return ans
        
    def findPath(self,root,target):
        stack = []
        lastVisited = None
        while root or stack:
            if root:
                stack.append(root)
                root = root.left
            else:
                peek = stack[-1]
                if peek.right and lastVisited != peek.right:
                    root = peek.right
                else:
                    if peek == target:
                        return stack
                    lastVisited = stack.pop()
                    root = None
        return stack

File: Tree/Lowest Common Ancestor of a Binary Tree/lowestCommonAncestor.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> findPath(TreeNode *root, TreeNode *p){
        vector<TreeNode*> path;
        TreeNode *lastVisited = NULL;
        while(root || !path.empty()){
            if(root){
                path.push_back(root);
                root = root->left;
            }
            else{
                TreeNode *node = path[path.size()-1];
                if(node->right && lastVisited!=node->right)
                    root = node->right;
                else{
                    if(node == p)
                        return path;
                    lastVisited = path[path.size()-1];
                    path.pop_back();
                    root = NULL;
                }
            }
        }
        return path;
    }
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        vector<TreeNode*> pathP = findPath(root,p), pathQ = findPath(root,q);
        TreeNode *ans = NULL;
        int x = 0;
        while(x < min(pathP.size(),pathQ.size()) && pathP[x] == pathQ[x]){
            ans = pathP[x];
            x++;
        }
        return ans;
    }
};
